# Importamos librerias

In [6]:
# Tratamiento de datos
# -----------------------------------------------------------------------
import pandas as pd
import numpy as np
import os


import sys 
sys.path.append("..")
from src import fx_limpieza as fx_l


# Configuración
# -----------------------------------------------------------------------
pd.set_option('display.max_columns', None) # para poder visualizar todas las columnas de los DataFrames

## Importamos todos los CSV y los nombramos con su respectivo año

In [7]:
df_13 = pd.read_csv("../datos/raw_data/datos-2013.csv", sep=';', encoding='latin-1')
df_14 = pd.read_csv("../datos/raw_data/datos-2014.csv", sep=';', encoding='latin-1')
df_15 = pd.read_csv("../datos/raw_data/datos-2015.csv", sep=';', encoding='latin-1')
df_16 = pd.read_csv("../datos/raw_data/datos-2016.csv", sep=';', encoding='latin-1')
df_17 = pd.read_csv("../datos/raw_data/datos-2017.csv", sep=';', encoding='latin-1')
df_18 = pd.read_csv("../datos/raw_data/datos-2018.csv", sep=';', encoding='latin-1')
df_19 = pd.read_csv("../datos/raw_data/datos-2019.csv", sep=';', encoding='latin-1')
df_20 = pd.read_csv("../datos/raw_data/datos-2020.csv", sep=';', encoding='latin-1')
df_21 = pd.read_csv("../datos/raw_data/datos-2021.csv", sep=';', encoding='latin-1')
diccionario = pd.read_csv("../datos/raw_data/diccionario_datos.csv", sep=';', encoding='latin-1')

In [8]:


# Ejemplo de uso
directorio = "../datos/raw_data"  # Especifica tu directorio
extension = '.csv'


df_resultado = fx_l.cargar_archivos_en_dataframe(directorio, extension,encoding="utf-8")

if df_resultado is not None:
    print(df_resultado.head())  # Muestra las primeras filas del DataFrame resultante


AttributeError: module 'src.fx_limpieza' has no attribute 'cargar_archivos_en_dataframe'

## Concatenamos todos los Dataframes

In [9]:
df = pd.concat([df_13, df_14, df_15,df_16, df_17, df_18, df_19, df_20, df_21], axis=0, ignore_index = True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1026299 entries, 0 to 1026298
Data columns (total 16 columns):
 #   Column                     Non-Null Count    Dtype 
---  ------                     --------------    ----- 
 0   CÓDIGO ÓRGÃO SUPERIOR      1026299 non-null  int64 
 1   NOME ÓRGÃO SUPERIOR        1026299 non-null  object
 2   CÓDIGO ÓRGÃO               1026299 non-null  int64 
 3   NOME ÓRGÃO                 1026299 non-null  object
 4   CÓDIGO UNIDADE GESTORA     1026299 non-null  int64 
 5   NOME UNIDADE GESTORA       1026299 non-null  object
 6   CATEGORIA ECONÔMICA        1026299 non-null  object
 7   ORIGEM RECEITA             1026299 non-null  object
 8   ESPÉCIE RECEITA            1026299 non-null  object
 9   DETALHAMENTO               1026299 non-null  object
 10  VALOR PREVISTO ATUALIZADO  1026299 non-null  object
 11  VALOR LANÇADO              1026299 non-null  object
 12  VALOR REALIZADO            1026299 non-null  object
 13  PERCENTUAL REALIZADO       

## Renombramos columnas para reducir barrera de lenguaje

In [10]:
df.rename(columns={
    "CÓDIGO ÓRGÃO SUPERIOR": "COD_organo_superior",
    "NOME ÓRGÃO SUPERIOR": "NOM_organo_superior",
    "CÓDIGO ÓRGÃO": "COD_organo",
    "NOME ÓRGÃO": "NOM_organo",
    "CÓDIGO UNIDADE GESTORA": "COD_unidad_gestora",
    "NOME UNIDADE GESTORA": "NOM_unidad_gestora",
    "CATEGORIA ECONÔMICA": "categoria_economica",
    "ORIGEM RECEITA": "origen_ingreso",
    "ESPÉCIE RECEITA": "especie_ingreso",
    "DETALHAMENTO": "detalle_adicional ",
    "VALOR PREVISTO ATUALIZADO": "valor_previsto_actualizado",
    "VALOR LANÇADO": "valor_lanzado",
    "VALOR REALIZADO": "valor_recaudado",
    "PERCENTUAL REALIZADO": "porcentaje_ejecutado",
    "DATA LANÇAMENTO": "fecha_ejecucion",
    "ANO EXERCÍCIO": "anio_ejecucion"
}, inplace=True)

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1026299 entries, 0 to 1026298
Data columns (total 16 columns):
 #   Column                      Non-Null Count    Dtype 
---  ------                      --------------    ----- 
 0   COD_organo_superior         1026299 non-null  int64 
 1   NOM_organo_superior         1026299 non-null  object
 2   COD_organo                  1026299 non-null  int64 
 3   NOM_organo                  1026299 non-null  object
 4   COD_unidad_gestora          1026299 non-null  int64 
 5   NOM_unidad_gestora          1026299 non-null  object
 6   categoria_economica         1026299 non-null  object
 7   origen_ingreso              1026299 non-null  object
 8   especie_ingreso             1026299 non-null  object
 9   detalle_adicional           1026299 non-null  object
 10  valor_previsto_actualizado  1026299 non-null  object
 11  valor_lanzado               1026299 non-null  object
 12  valor_recaudado             1026299 non-null  object
 13  porcentaje_e

## Pasamos todos los valores que tenemos como objeto a formato numero

In [12]:
df["valor_previsto_actualizado"] = df["valor_previsto_actualizado"].str.replace(",", ".").astype(float)

In [13]:
df["valor_lanzado"] = df["valor_lanzado"].str.replace(",", ".").astype(float)

In [14]:
df["valor_recaudado"] = df["valor_recaudado"].str.replace(",", ".").astype(float)

In [15]:
df["porcentaje_ejecutado"] = df["porcentaje_ejecutado"].str.replace(",", ".").astype(float)

## Pasamos los valores que tenemos como objeto a formato fecha


In [16]:
# Convertimos la columna a formato datetime
df["fecha_ejecucion"] = pd.to_datetime(df["fecha_ejecucion"], errors="coerce", dayfirst="True")

## Chequeamos los valores nulos por columnas

In [17]:
## Llamamos a la funcion de nuestra biblioteca de funciones.
fx_l.valores_nulos(df)

numero_nulos  porcentaje_nulos       tipo_dato
COD_organo_superior                    0              0.00           int64
NOM_organo_superior                    0              0.00          object
COD_organo                             0              0.00           int64
NOM_organo                             0              0.00          object
COD_unidad_gestora                     0              0.00           int64
NOM_unidad_gestora                     0              0.00          object
categoria_economica                    0              0.00          object
origen_ingreso                         0              0.00          object
especie_ingreso                        0              0.00          object
detalle_adicional                      0              0.00          object
valor_previsto_actualizado             0              0.00         float64
valor_lanzado                          0              0.00         float64
valor_recaudado                        0              0.00         float64
porcentaje_ejecutado                   0              0.00         float64
fecha_ejecucion                      578              0.06  datetime64[ns]
anio_ejecucion                         0              0.00           int64

#### Solamente tenemos nulos en la columna de fecha de ejecucion. Lo que se puede hacer es reemplazarlos con el año que tenemos en la columna de al lado y ponerle 01-01-año

#### Igualmente vemos que muchas columnas relacionadas con los valores o float tienen como valor 0

In [18]:
df.sample(10)

COD_organo_superior                            NOM_organo_superior  \
481650                 26000                         Ministério da Educação   
29615                  24000  Ministério da Ciência, Tecnologia, Inovações    
669774                 26000                         Ministério da Educação   
915135                 52000                           Ministério da Defesa   
1001137                30000      Ministério da Justiça e Segurança Pública   
260903                 25000                         Ministério da Economia   
73736                  25000                         Ministério da Economia   
679971                 26000                         Ministério da Educação   
570182                 54000                          Ministério do Turismo   
15931                  22000  Ministério da Agricultura, Pecuária e Abastec   

         COD_organo                                         NOM_organo  \
481650        26247                Universidade Federal de Santa Maria   
29615         24901  Fundo Nacional de Desenvolvimento Científico e...   
669774        26278           Fundação Universidade Federal de Pelotas   
915135        52911                                  Fundo Aeronáutico   
1001137       30907                       Fundo Penitenciário Nacional   
260903        25000  Ministério da Economia - Unidades com vínculo ...   
73736         25000  Ministério da Economia - Unidades com vínculo ...   
679971        26291  Fundação Coordenação de Aperfeiçoamento de Pes...   
570182        34902                          Fundo Nacional de Cultura   
15931         22202        Empresa Brasileira de Pesquisa Agropecuária   

         COD_unidad_gestora                             NOM_unidad_gestora  \
481650               153164            UNIVERSIDADE FEDERAL DE SANTA MARIA   
29615                240901      FUNDO NAC.DE DESENV. CIENT. E TECNOLOGICO   
669774               154047       FUNDACAO UNIVERSIDADE FEDERAL DE PELOTAS   
915135               121002      SECRETARIA DE ECON. E FIN.DA AER.- F.AER.   
1001137              200333                    DEPEN - DIRETORIA EXECUTIVA   
260903               170013        SETORIAL ORCAMENTARIA E FINANCEIRA / ME   
73736                170013        SETORIAL ORCAMENTARIA E FINANCEIRA / ME   
679971               154003  FUND.COORD.DE APERF.DE PESSOAL NIVEL SUPERIOR   
570182               340002    COORD.GERAL DE PLAN. ORC. FIN. E CONTAB/FNC   
15931                135037                               EMBRAPA-SETORIAL   

        categoria_economica                               origen_ingreso  \
481650   Receitas Correntes                          Receita de Serviços   
29615    Receitas Correntes                                Contribuições   
669774   Receitas Correntes                    Outras Receitas Correntes   
915135   Receitas Correntes                          Receita de Serviços   
1001137  Receitas Correntes                    Outras Receitas Correntes   
260903   Receitas Correntes  Impostos, Taxas e Contribuições de Melhoria   
73736    Receitas Correntes                                Contribuições   
679971   Receitas Correntes                    Outras Receitas Correntes   
570182   Receitas Correntes                    Outras Receitas Correntes   
15931    Receitas Correntes                         Receita Agropecuária   

                                       especie_ingreso  \
481650    Serviços Administrativos e Comerciais Gerais   
29615                         Contribuições econômicas   
669774     Indenizações, restituições e ressarcimentos   
915135   Serviços e Atividades Referentes à Navegação    
1001137  Bens, Direitos e Valores Incorporados ao Patr   
260903                                        Impostos   
73736                            Contribuições sociais   
679971     Indenizações, restituições e ressarcimentos   
570182   Multas administrativas, contratuais e judicia   
15931                             Receita Agrope

In [19]:
# Reemplaza los nulos y convierte a fecha
df["fecha_ejecucion"] = df["fecha_ejecucion"].fillna(
    df["anio_ejecucion"].astype(str)+"-01-01"  # Supone que el formato es YYYY-MM-DD
)


In [20]:
df.sample()

COD_organo_superior            NOM_organo_superior  COD_organo  \
185481                32000  Ministério de Minas e Energia       32210   

                                  NOM_organo  COD_unidad_gestora  \
185481  Agência Nacional de Energia Elétrica              323028   

                          NOM_unidad_gestora categoria_economica  \
185481  AGENCIA NACIONAL DE ENERGIA ELETRICA  Receitas Correntes   

             origen_ingreso                                especie_ingreso  \
185481  Receita Patrimonial  Exploração do patrimônio imobiliário do Estad   

                                  detalle_adicional   \
185481  CONC/PERM/AUT/CESSAO DIR.USO IMOV.PUB-PRINC.   

        valor_previsto_actualizado  valor_lanzado  valor_recaudado  \
185481                         0.0            0.0      49382463.66   

        porcentaje_ejecutado fecha_ejecucion  anio_ejecucion  
185481                   0.0      2016-01-15            2016

In [21]:
#Comprobamos que hayan desaparecido los nulos
print("Null values per column:")
display(df.isna().sum())

Null values per column:


COD_organo_superior           0
NOM_organo_superior           0
COD_organo                    0
NOM_organo                    0
COD_unidad_gestora            0
NOM_unidad_gestora            0
categoria_economica           0
origen_ingreso                0
especie_ingreso               0
detalle_adicional             0
valor_previsto_actualizado    0
valor_lanzado                 0
valor_recaudado               0
porcentaje_ejecutado          0
fecha_ejecucion               0
anio_ejecucion                0
dtype: int64

## Tratamiento de duplicados

In [22]:
duplicados = df[df.duplicated()]
duplicados

Empty DataFrame
Columns: [COD_organo_superior, NOM_organo_superior, COD_organo, NOM_organo, COD_unidad_gestora, NOM_unidad_gestora, categoria_economica, origen_ingreso, especie_ingreso, detalle_adicional , valor_previsto_actualizado, valor_lanzado, valor_recaudado, porcentaje_ejecutado, fecha_ejecucion, anio_ejecucion]
Index: []

#### No se encontraron valores duplicados en nuestro dataframe

**Insights:** 

- El DataFrame contiene 1026299 registros y 16 columnas con información sobre características de recaudacion y ejecucion de presupuestos de diferentes ministerios

- Despues de la limpieza de datos la columnas tienen datos completos 

- Los tipos de datos incluyen 4 columnas numéricas enteras (int64), 7 categóricas (object), 4 numérica flotante (float64) y 1 de fecha (datetime64[ns]). 

- El DataFrame ocupa aproximadamente 125.3+ KB en memoria.

In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1026299 entries, 0 to 1026298
Data columns (total 16 columns):
 #   Column                      Non-Null Count    Dtype         
---  ------                      --------------    -----         
 0   COD_organo_superior         1026299 non-null  int64         
 1   NOM_organo_superior         1026299 non-null  object        
 2   COD_organo                  1026299 non-null  int64         
 3   NOM_organo                  1026299 non-null  object        
 4   COD_unidad_gestora          1026299 non-null  int64         
 5   NOM_unidad_gestora          1026299 non-null  object        
 6   categoria_economica         1026299 non-null  object        
 7   origen_ingreso              1026299 non-null  object        
 8   especie_ingreso             1026299 non-null  object        
 9   detalle_adicional           1026299 non-null  object        
 10  valor_previsto_actualizado  1026299 non-null  float64       
 11  valor_lanzado           

# Finalizada la limpieza de datos, exportamos el df limpio

In [24]:
df.to_pickle("../datos/output_data/df_limpio.pkl")